java/vector/src/main/codegen/templates/DenseUnionVector.java

```diff
@@ -714,8 +714,22 @@ public int getBufferSizeFor(final int count) {
     if (count == 0) {
       return 0;
     }
+
+    int[] legCounts = new int[typeFields.length];
+    for (int i = 0; i < count; i++) {
+      byte tid = getTypeId(i);
+      legCounts[tid]++;
+    }
+
+    long legBytes = 0;
+    byte tid = 0;
+    for (ValueVector v : internalStruct) {
+      legBytes += (long)v.getBufferSizeFor(legCounts[tid]);
+      tid++;
+    }
```

Comment:
I don't think this is right: `internalStruct` has its fields indexed by physical child id (that is: 0, 1, 2...) but the type ids are logical values that are unrelated to the physical structure.

But I think that you should just copy what is already done in `setChildVectorValueCounts` and adapt it for the use case of counting buffer bytes.